File: CustomApplications/VB.Net/OnAnimUpdate/Form1.vb
Imports System.Math
Imports System.Runtime.InteropServices


Public Class OnAnimUpdate
    Private WithEvents STKXApp As New AGI.STKX.AgSTKXApplication
    Private oldX As Single
    Private oldY As Single
    Private stkRootObject As AGI.STKObjects.AgStkObjectRoot = Nothing
    Private oldButton As Integer

    Public Shared Sub Main()
        Dim STKXApp As AGI.STKX.AgSTKXApplication = Nothing
        Try

            STKXApp = New AGI.STKX.AgSTKXApplication()
            If (Not STKXApp.IsFeatureAvailable(AGI.STKX.AgEFeatureCodes.eFeatureCodeGlobeControl)) Then
                MessageBox.Show("You do not have the required license.", "License Error", MessageBoxButtons.OK, MessageBoxIcon.Stop)
            End If

        Catch exception As System.Runtime.InteropServices.COMException

            If (exception.ErrorCode = &H80040154) Then

                Dim errorMessage As String = "Could not instantiate AgSTKXApplication."
                errorMessage += Environment.NewLine
                errorMessage += Environment.NewLine
                errorMessage += "You are trying to run in the x64 configuration. Check that STK Engine 64-bit is installed on this machine."

                MessageBox.Show(errorMessage, "STK Engine Error", MessageBoxButtons.OK, MessageBoxIcon.Stop)

            Else
                Throw
            End If

        End Try

        If Not STKXApp Is Nothing Then
            Application.Run(New OnAnimUpdate)
        End If
    End Sub

    Private ReadOnly Property stkRoot() As AGI.STKObjects.AgStkObjectRoot
        Get
            If stkRootObject Is Nothing Then
                stkRootObject = New AGI.STKObjects.AgStkObjectRoot
            End If
            stkRoot = stkRootObject
        End Get
    End Property


    Private Sub OnAnimeUpdate_Load(ByVal sender As Object, ByVal e As System.EventArgs) Handles MyBase.Load
        Me.StatusBar1.Panels.Add("Panel 1")
        Me.StatusBar1.Panels.Add("Panel 2")
        Me.StatusBar1.Panels.Add("Panel 3")
        Me.StatusBar1.Panels(0).Width = Me.StatusBar1.Width / 3
        Me.StatusBar1.Panels(1).Width = Me.StatusBar1.Width / 3
        Me.StatusBar1.Panels(2).Width = Me.StatusBar1.Width / 3       
        stkRoot.NewScenario("Test")
        stkRoot.ExecuteCommand("ConControl / VerboseOff")
    End Sub

    Private Sub MenuItem1_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MenuItem1.Click
        OpenScenario.ShowDialog()
    End Sub

    Private Sub AxAgUiAxVOCntrl1_MouseUpEvent(ByVal sender As Object, ByVal e As AxAGI.STKX.IAgUiAxVOCntrlEvents_MouseUpEvent) Handles AxAgUiAxVOCntrl1.MouseUpEvent
        DoMouseUp(e.button, e.shift, e.x, e.y)
    End Sub

    Private Sub MenuItem4_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MenuItem4.Click
        stkRoot.PlayForward()
    End Sub

    Private Sub MenuItem5_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MenuItem5.Click
        stkRoot.Pause()
    End Sub

    Private Sub MenuItem6_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MenuItem6.Click
        stkRoot.Rewind()
    End Sub

    Private Sub MenuItem7_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MenuItem7.Click
        stkRoot.Faster()
    End Sub

    Private Sub MenuItem8_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MenuItem8.Click
        stkRoot.Slower()
    End Sub

    Private Sub MenuItem9_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MenuItem9.Click
        Me.AxAgUiAxVOCntrl1.ZoomIn()
    End Sub

    Private Sub MenuItem10_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MenuItem10.Click
        stkRoot.ExecuteCommand("VO * View Home")
    End Sub

    Private Sub OpenScenario_FileOk(ByVal sender As Object, ByVal e As System.ComponentModel.CancelEventArgs) Handles OpenScenario.FileOk
        stkRoot.CloseScenario()
        stkRoot.LoadScenario(OpenScenario.FileName)
    End Sub

    Private Sub STKXApp_OnAnimUpdate(ByVal TimeEpSec As Double) Handles STKXApp.OnAnimUpdate
        Me.StatusBar1.Panels(0).Text = CStr(TimeEpSec) & " EpSec"

        Dim retList As AGI.STKUtil.AgExecCmdResult
        retList = stkRoot.ExecuteCommand("GetAnimTime *")

        If retList.Count > 0 Then
            Me.StatusBar1.Panels(2).Text = retList(0)
        Else
            Me.StatusBar1.Panels(2).Text = "N/A"
        End If

        retList = stkRoot.ExecuteCommand("AnimFrameRate *")

        If retList.Count > 0 And retList(0) > 0 Then
            Me.StatusBar1.Panels(1).Text = retList(0) & " fps"
        Else
            Me.StatusBar1.Panels(1).Text = "N/A"
        End If
    End Sub
    Private Sub DoMouseDown(ByVal Button As Integer, ByVal Shift As Integer, ByVal x As Single, ByVal y As Single)
        oldX = x
        oldY = y
        oldButton = Button
    End Sub
    Private Sub DoMouseUp(ByVal Button As Integer, ByVal Shift As Integer, ByVal x As Single, ByVal y As Single)
        If oldButton = 2 And Shift = 0 And Abs(oldX - x) < 10 And Abs(oldY - y) < 10 Then
            Dim r As System.Drawing.Point
            r.X = x
            r.Y = y
            PopupMenu.Show(Me, r)
        End If
    End Sub

    Private Sub AxAgUiAxVOCntrl1_MouseDownEvent(ByVal sender As Object, ByVal e As AxAGI.STKX.IAgUiAxVOCntrlEvents_MouseDownEvent) Handles AxAgUiAxVOCntrl1.MouseDownEvent
        DoMouseDown(e.button, e.shift, e.x, e.y)
    End Sub

    Private Sub OnAnimUpdate_Resize(ByVal sender As Object, ByVal e As System.EventArgs) Handles MyBase.Resize
        Me.AxAgUiAxVOCntrl1.Left = 0
        Me.AxAgUiAxVOCntrl1.Top = 0
        Me.AxAgUiAxVOCntrl1.Width = Me.ClientRectangle.Width
        Me.AxAgUiAxVOCntrl1.Height = Me.ClientRectangle.Height - Me.StatusBar1.Height
    End Sub

    Private Sub OnAnimUpdate_FormClosing(ByVal sender As System.Object, ByVal e As System.Windows.Forms.FormClosingEventArgs) Handles MyBase.FormClosing
        If Not (IsNothing(stkRootObject)) Then
            stkRootObject.CloseScenario()
        End If
    End Sub
End Class

File: CustomApplications/CSharp/OnAnimUpdate/Form1.cs
using System;
using System.Drawing;
using System.Windows.Forms;



namespace OnAnimUpdate
{
	/// <summary>
	/// Summary description for Form1.
	/// </summary>
	public partial class Form1 : System.Windows.Forms.Form
	{
		private AGI.STKX.AgSTKXApplication STKXApp;
        private AGI.STKObjects.AgStkObjectRoot stkRootObject = null;

		private int oldX;
		private int oldY;

		private int oldButton;

        private AGI.STKObjects.AgStkObjectRoot stkRoot
        {
            get
            {
                if (stkRootObject == null)
                {
                    stkRootObject = new AGI.STKObjects.AgStkObjectRoot();
                }
                return stkRootObject;
            }
        }

		public Form1()
		{
			//
			// Required for Windows Form Designer support
			//
			InitializeComponent();

			//
			// TODO: Add any constructor code after InitializeComponent call
			//
			this.STKXApp = new AGI.STKX.AgSTKXApplication();
			this.STKXApp.OnAnimUpdate += this.STKXApp_OnAnimUpdate;			

			StatusBarPanel pn10;
			StatusBarPanel pn11;
			StatusBarPanel pn12;
			
			pn10 = new StatusBarPanel();
			pn11 = new StatusBarPanel();
			pn12 = new StatusBarPanel();

			this.statusBar1.Panels.Add(pn10);
			this.statusBar1.Panels.Add(pn11);
			this.statusBar1.Panels.Add(pn12);
		}

		/// <summary>
		/// The main entry point for the application.
		/// </summary>
		[STAThread]
		static void Main() 
		{
            AGI.STKX.AgSTKXApplication STKXApp = null;
            try
            {
                STKXApp = new AGI.STKX.AgSTKXApplication();

                if (!STKXApp.IsFeatureAvailable(AGI.STKX.AgEFeatureCodes.eFeatureCodeGlobeControl))
                {
                    MessageBox.Show("You do not have the required license.", "License Error", MessageBoxButtons.OK, MessageBoxIcon.Stop);
                }
            }
            catch (System.Runtime.InteropServices.COMException exception)
            {
                if (exception.ErrorCode == unchecked((int)0x80040154))
                {
                    string errorMessage = "Could not instantiate AgSTKXApplication.";
                    errorMessage += Environment.NewLine;
                    errorMessage += Environment.NewLine;
                    errorMessage += "Check that STK or STK Engine 64-bit is installed on this machine.";

                    MessageBox.Show(errorMessage, "STK Engine Error", MessageBoxButtons.OK, MessageBoxIcon.Stop);
                }
                else
                {
                    throw;
                }
            }
            if (STKXApp != null)
            {
                Application.Run(new Form1());
            }
		}
		
		private void Form1_Load(object sender, System.EventArgs e)
		{
			StatusBarPanel pn10 = this.statusBar1.Panels[0];
			StatusBarPanel pn11 = this.statusBar1.Panels[1];
			StatusBarPanel pn12 = this.statusBar1.Panels[2];

			pn10.Width=this.Width/3;
			pn11.Width=this.Width/3;
			pn12.Width=this.Width/3;

			stkRoot.NewScenario("Test");
			stkRoot.ExecuteCommand("ConControl / VerboseOff");
		}

		private string FilterChars( string text )
		{
			string result;

			result=text.Replace("\r","");
			result=text.Replace("\n","");
			result=text.Replace("\"","");

			return (result);
		}

		private void Form1_Resize(object sender, System.EventArgs e)
		{
			this.axAgUiAxVOCntrl1.Left=0;
			this.axAgUiAxVOCntrl1.Top=0;
			this.axAgUiAxVOCntrl1.Width=this.ClientRectangle.Width;
			this.axAgUiAxVOCntrl1.Height=this.ClientRectangle.Height-this.statusBar1.Height;
			
			StatusBarPanel pn10 = this.statusBar1.Panels[0];
			StatusBarPanel pn11 = this.statusBar1.Panels[1];
			StatusBarPanel pn12 = this.statusBar1.Panels[2];
			
			pn10.Width = this.Width / 3;
			pn11.Width = this.Width / 3;
			pn12.Width = this.Width / 3;
		
		}

		private void DoMouseDown(int Button, int Shift, int x, int y)
		{
			oldX = x;
			oldY = y;
			oldButton = Button;
		}

		private void DoMouseUp(int Button, int Shift, int x, int y)
		{
			if (oldButton==2 & Shift==0 & Math.Abs(oldX-x)<10 & Math.Abs(oldY-y)<10)
			{
				this.MenuRoot.Show(this, new Point(x,y));
			}
		}

		private void MenuAnimFaster_Click(object sender, System.EventArgs e)
		{
			stkRoot.Faster();
		}

		private void MenuAnimPause_Click(object sender, System.EventArgs e)
		{
			stkRoot.Pause();
		}

		private void MenuAnimReset_Click(object sender, System.EventArgs e)
		{
			stkRoot.Rewind();
		}

		private void MenuAnimSlower_Click(object sender, System.EventArgs e)
		{
			stkRoot.Slower();
		}

		private void MenuAnimStart_Click(object sender, System.EventArgs e)
		{
			stkRoot.PlayForward();
		}

		private void MenuHomeView_Click(object sender, System.EventArgs e)
		{
			stkRoot.ExecuteCommand("VO * View Home");
		}

		private void MenuZoomIn_Click(object sender, System.EventArgs e)
		{
			this.axAgUiAxVOCntrl1.ZoomIn();
		}

		private void axAgUiAxVOCntrl1_MouseDownEvent(object sender, AxAGI.STKX.IAgUiAxVOCntrlEvents_MouseDownEvent e)
		{
			DoMouseDown(e.button, e.shift, e.x, e.y);
		}

		private void axAgUiAxVOCntrl1_MouseUpEvent(object sender, AxAGI.STKX.IAgUiAxVOCntrlEvents_MouseUpEvent e)
		{
			DoMouseUp(e.button, e.shift, e.x, e.y);
		}

		private void MenuLoadScenario_Click(object sender, System.EventArgs e)
		{
			openFileDialog1.InitialDirectory = @"C:\My Documents" ;
			openFileDialog1.Filter = "Scenario (.sc)|*.sc" ;
			openFileDialog1.FilterIndex = 2 ;
			openFileDialog1.RestoreDirectory = true ;

			if(openFileDialog1.ShowDialog() == DialogResult.OK)
			{
				stkRoot.CloseScenario();
				stkRoot.LoadScenario(this.openFileDialog1.FileName);
			}

		}

		private void STKXApp_OnAnimUpdate(double TimeEpSec)
		{
			this.statusBar1.Panels[0].Text = TimeEpSec + " EpSec";
			AGI.STKUtil.AgExecCmdResult retlist;
            retlist = (AGI.STKUtil.AgExecCmdResult)stkRoot.ExecuteCommand("GetAnimTime *");

			if ( retlist.Count > 0 )
				this.statusBar1.Panels[1].Text = FilterChars(retlist[0]);
			else
 				this.statusBar1.Panels[1].Text = "N/A";

            retlist = (AGI.STKUtil.AgExecCmdResult)stkRoot.ExecuteCommand("AnimFrameRate *");

			if ( retlist.Count > 0 & retlist[0] != null )
				this.statusBar1.Panels[2].Text = FilterChars( retlist[0] ) + " fps";
			else
				this.statusBar1.Panels[2].Text = "N/A";

		}

		private void Form1_Closed(object sender, System.EventArgs e)
		{
			stkRoot.Pause();
		}

		private void axAgUiAxVOCntrl1_OLEDragDrop(object sender, AxAGI.STKX.IAgUiAxVOCntrlEvents_OLEDragDropEvent e)
		{
		
		}

        private void Form1_FormClosing(object sender, FormClosingEventArgs e)
        {
            if (stkRootObject != null)
            {
                stkRootObject.CloseScenario();
            }
        }

	}
}
